File: java-worker/src/MLModel.java
public class MLModel {
    private String modelId;
    private double[] weights;
    private double bias;
    private int inputSize;
    
    public MLModel(String modelId) {
        this.modelId = modelId;
    }
    
    public void train(String inputData, String outputData) {
        // Implementación simple de regresión lineal
        String[] inputs = inputData.split(",");
        String[] outputs = outputData.split(",");
          // Determinar el tamaño de entrada - flexible con separadores
        if (inputs[0].contains(";")) {
            inputSize = inputs[0].split(";").length;
        } else if (inputs[0].contains(",")) {
            inputSize = inputs[0].split(",").length;
        } else {
            inputSize = 1; // Entrada simple
        }
        
        weights = new double[inputSize];
        
        // Gradiente descendente simple
        double learningRate = 0.01;
        int epochs = 100;
        
        for (int epoch = 0; epoch < epochs; epoch++) {
            for (int i = 0; i < inputs.length; i++) {
                double[] x = parseInput(inputs[i]);
                double y = Double.parseDouble(outputs[i]);
                
                double prediction = predict(x);
                double error = y - prediction;
                
                // Actualizar pesos
                for (int j = 0; j < weights.length; j++) {
                    weights[j] += learningRate * error * x[j];
                }
                bias += learningRate * error;
            }
        }
    }
    
    public String predict(String inputData) {
        double[] x = parseInput(inputData);
        return String.valueOf(predict(x));
    }
    
    private double predict(double[] x) {
        double result = bias;
        for (int i = 0; i < weights.length && i < x.length; i++) {
            result += weights[i] * x[i];
        }
        return result;
    }
      private double[] parseInput(String input) {
        // Ser flexible: aceptar tanto ; como , como separador
        String[] parts;
        if (input.contains(";")) {
            parts = input.split(";");
        } else if (input.contains(",")) {
            parts = input.split(",");
        } else {
            // Un solo valor
            return new double[]{Double.parseDouble(input.trim())};
        }
        
        double[] result = new double[parts.length];
        for (int i = 0; i < parts.length; i++) {
            result[i] = Double.parseDouble(parts[i].trim());
        }
        return result;
    }
    
    public String serialize() {
        StringBuilder sb = new StringBuilder();
        sb.append(inputSize).append("|");
        sb.append(bias).append("|");
        for (double w : weights) {
            sb.append(w).append(";");
        }
        return sb.toString();
    }
    
    public void deserialize(String data) {
        String[] parts = data.split("\\|");
        if (parts.length >= 3) {
            inputSize = Integer.parseInt(parts[0]);
            bias = Double.parseDouble(parts[1]);
            String[] weightParts = parts[2].split(";");
            weights = new double[inputSize];
            for (int i = 0; i < inputSize && i < weightParts.length; i++) {
                if (!weightParts[i].isEmpty()) {
                    weights[i] = Double.parseDouble(weightParts[i]);
                }
            }
        }
    }
}

File: python-worker/ml_model.py
class MLModel:
    def __init__(self, model_id):
        self.model_id = model_id
        self.weights = []
        self.bias = 0.0
        self.input_size = 0
    
    def train(self, input_data, output_data):
        print(f"DEBUG: Iniciando entrenamiento con input_data={input_data}, output_data={output_data}")
        
        # Parse training data
        inputs = input_data.split(',')
        outputs = output_data.split(',')
        
        if not inputs or not outputs:
            raise ValueError("Datos de entrenamiento vacíos")
        
        if len(inputs) != len(outputs):
            raise ValueError(f"Cantidad de entradas ({len(inputs)}) no coincide con salidas ({len(outputs)})")
        
        # Determinar el formato de entrada del primer ejemplo
        first_input = inputs[0].strip()
        if ';' in first_input:
            # Formato múltiple: "1;2"
            self.input_size = len(first_input.split(';'))
        else:
            # Formato simple: "1"
            self.input_size = 1
        
        self.weights = [0.0] * self.input_size
        
        print(f"DEBUG: Input size detectado: {self.input_size}")
        
        # Simple gradient descent
        learning_rate = 0.01
        epochs = 100
        
        for epoch in range(epochs):
            total_error = 0
            for i in range(len(inputs)):
                try:
                    x = self._parse_input(inputs[i])
                    y = float(outputs[i])
                    
                    # Forward pass
                    prediction = self._predict(x)
                    error = y - prediction
                    total_error += error ** 2
                    
                    # Update weights
                    for j in range(len(self.weights)):
                        if j < len(x):  # Verificar índice
                            self.weights[j] += learning_rate * error * x[j]
                    self.bias += learning_rate * error
                except Exception as e:
                    print(f"Error en epoch {epoch}, ejemplo {i}: {e}")
                    raise
        
        print(f"DEBUG: Entrenamiento completado. Weights: {self.weights}, Bias: {self.bias}")
    
    def predict(self, input_data):
        x = self._parse_input(input_data)
        prediction = self._predict(x)
        return str(prediction)
    
    def _predict(self, x):
        result = self.bias
        for i in range(min(len(self.weights), len(x))):
            result += self.weights[i] * x[i]
        return result
    
    def _parse_input(self, input_str):
        input_str = input_str.strip()
        
        # Manejar diferentes formatos
        if ';' in input_str:
            # Formato múltiple: "1;2"
            parts = input_str.split(';')
        else:
            # Formato simple: "1" 
            parts = [input_str]
        
        # Convertir a float y validar
        result = []
        for part in parts:
            try:
                result.append(float(part.strip()))
            except ValueError:
                raise ValueError(f"No se puede convertir '{part}' a número")
        
        # Asegurar que el tamaño coincida con lo esperado
        if self.input_size > 0 and len(result) != self.input_size:
            raise ValueError(f"Entrada tiene {len(result)} valores, se esperaban {self.input_size}")
        
        return result
    
    def serialize(self):
        weights_str = ';'.join(str(w) for w in self.weights)
        return f"{self.input_size}|{self.bias}|{weights_str}"
    
    def deserialize(self, data):
        parts = data.split('|')
        if len(parts) >= 3:
            self.input_size = int(parts[0])
            self.bias = float(parts[1])
            weight_parts = parts[2].split(';')
            self.weights = []
            for w in weight_parts:
                if w:  # Ignorar strings vacíos
                    self.weights.append(float(w))